app/services/move_work_package_service.rb

```diff
@@ -0,0 +1,191 @@
+
+# Moves/copies an work_package to a new project and type
+# Returns the moved/copied work_package on success, false on failure
+
+class MoveWorkPackageService
+  attr_accessor :work_package,
+                :user
+
+  def initialize(work_package, user)
+    self.work_package = work_package
+    self.user = user
+  end
+
+  def call(new_project, new_type = nil, options = {})
+    if options[:no_transaction]
+      move_without_transaction(new_project, new_type, options)
+    else
+      WorkPackage.transaction do
+        move_without_transaction(new_project, new_type, options) ||
+          raise(ActiveRecord::Rollback)
+      end || false
+    end
+  end
+
+  private
+
+  def move_without_transaction(new_project, new_type = nil, options = {})
+    attributes = options[:attributes] || {}
+
+    modified_work_package = copy_or_move(options[:copy], new_project, new_type, attributes)
+
+    if options[:copy]
+      return false unless copy(modified_work_package, attributes, options)
+    else
+      return false unless move(modified_work_package, new_project, options)
+    end
+
+    modified_work_package
+  end
+
+  def copy_or_move(make_copy, new_project, new_type, attributes)
+    modified_work_package = if make_copy
+                              WorkPackage.new.copy_from(work_package)
+                            else
+                              work_package
+                            end
+
+    move_to_project(modified_work_package, new_project)
+
+    move_to_type(modified_work_package, new_type)
+
+    bulk_assign_attributes(modified_work_package, attributes)
+
+    modified_work_package
+  end
+
+  def copy(modified_work_package, attributes, options)
+    set_default_values_on_copy(modified_work_package, attributes)
+
+    return false unless modified_work_package.save
+
+    create_and_save_journal_note modified_work_package, options[:journal_note]
+
+    true
+  end
+
+  def move(modified_work_package, new_project, options)
+    if options[:journal_note]
+      modified_work_package.add_journal user, options[:journal_note]
+    end
+
+    return false unless modified_work_package.save
+
+    move_time_entries(modified_work_package, new_project)
+
+    return false unless move_children(modified_work_package, new_project, options)
+
+    true
```

Comment:
There is a subtle difference in behaviour when I change it to what you propose:

`return move_children` will always return. But I want to only return if move_children fails.

Granted, I could combine it with the implicit `return true` below, but I thought that it would read more consistent with the code above. Something along the line of "you can fail up until the last statement but if you are there, signal true".